src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToForEachProvider.cs

```diff
@@ -0,0 +1,769 @@
+                    prefix,
+                    _semanticModel.LookupSymbols(position).Select(symbol => symbol.Name).ToImmutableHashSet(),
+                    StringComparer.CurrentCulture);
+
+            private bool TryConvertIfInForEach(QueryExpressionSyntax source, ForEachStatementSyntax forEachStatement, out DocumentUpdateInfo documentUpdateInfo)
+            {
+                // before foreach(var x in from a in b select a)
+                var statement = forEachStatement.Statement;
+                StatementSyntax internalNodeMethod(ExpressionSyntax expression)
+                {
+                    // before foreach(var a in from ... a) { dosomething(a); }
+                    // after 
+                    // foreach (var a in ...)
+                    // ...
+                    //{ dosomething(a); }
+                    if (expression is IdentifierNameSyntax identifier &&
+                        identifier.Identifier.ValueText == forEachStatement.Identifier.ValueText)
+                    {
+                        return statement;
+                    }
+                    else
+                    {
+                        // before foreach(var x in from ... a) { dosomething(x); }
+                        // after 
+                        // foreach (var a in ...)
+                        // ...
+                        // {
+                        //      var x = a;
+                        //      dosomething(x); 
+                        //  }
+                        var declaration = SyntaxFactory.LocalDeclarationStatement(
+                            SyntaxFactory.VariableDeclaration(
+                                VarNameIdentifier,
+                                SyntaxFactory.SingletonSeparatedList(
+                                    SyntaxFactory.VariableDeclarator(
+                                        forEachStatement.Identifier,
+                                        argumentList: null,
+                                        SyntaxFactory.EqualsValueClause(expression)))));
+                        return AddToBlockTop(declaration, statement);
+                    }
+                }
+
+                if (TryConvertQueryExpression(source, internalNodeMethod, out StatementSyntax newNode))
+                {
+                    documentUpdateInfo = new DocumentUpdateInfo(forEachStatement, newNode);
+                    return true;
+                }
+
+                documentUpdateInfo = default;
+                return false;
+            }
+
+            private bool TryConvertIfInReturnStatement(QueryExpressionSyntax source, ReturnStatementSyntax returnStatement, out DocumentUpdateInfo documentUpdateInfo)
+            {
+                // before: return from a in b select a;
+                // after: 
+                // foreach(var a in b)
+                // {
+                //      yield return a;
+                // }
```

Comment:
this is going to be a bit wonky in that other parts of hte method that previously did 'returns' will now be broken.  But i suppose that's ok.